File: backend/app/services/orchestrator.py
"""
Orchestrator — 3-phase pipeline running in background thread.
Phase 1: all specialist agents write to SharedGraph concurrently
Phase 2: synthesizer(s) read complete graph -> verdict
Phase 3: optional super-synthesizer reconciles verdicts
"""
import time, uuid
from concurrent.futures import ThreadPoolExecutor, as_completed
from .data_collector import build_world_state, serialize
from .graph_builder import SharedGraph
from .agent_runner import run_agent, run_synthesizer, run_super_synthesizer
from .relevance import get_relevance_context
from ..models.task import TaskManager, TaskStatus
from ..utils.logger import get_logger
log = get_logger('assetflow.orchestrator')

MAX_W = 5

def _emit(task, status=None, progress=None, message=None, result=None, error=None):
    kw = {}
    if status   is not None: kw['status']   = status
    if progress is not None: kw['progress'] = progress
    if message  is not None: kw['message']  = message
    if result   is not None: kw['result']   = result
    if error    is not None: kw['error']    = error
    TaskManager.update(task.task_id, **kw)

def run_pipeline(asset: dict, agents: list, task, store, on_complete=None):
    t0 = time.time()
    analysis_id = f"an_{uuid.uuid4().hex[:10]}"
    try:
        enabled = [a for a in agents if a.enabled]
        if not enabled:
            raise ValueError("No enabled agents.")

        super_agents = [a for a in enabled if a.role == 'super_synthesizer']
        synth_agents  = [a for a in enabled if a.role == 'synthesizer']
        spec_agents   = [a for a in enabled if a.role not in ('synthesizer','super_synthesizer')]
        p1 = spec_agents or enabled
        synths = synth_agents or [p1[0]]
        super_a = super_agents[0] if super_agents else None

        def prog(pct, msg):
            _emit(task, status=TaskStatus.RUNNING, progress=pct, message=msg)
            log.info(f"[{task.task_id}] {pct}% - {msg}")

        # PHASE 0: FETCH ALL WORLD DATA
        prog(5, 'Fetching world data: price, news, weather, social, commodities...')
        extra_texts = asset.pop('extra_texts', [])
        ws = None
        try:
            ws = build_world_state(asset['symbol'], asset.get('av_key'), extra_texts)
        except Exception as e:
            log.warning(f'World state partial failure: {e}')

        weather_alerts = sum(1 for w in (ws or {}).get('weather',[]) if w.get('significant'))
        social_total   = (ws or {}).get('social',{}).get('stats',{}).get('total',0)
        prog(20, f"Data ready | {weather_alerts} weather alerts | {social_total} social posts")

        rel_ctx = get_relevance_context(asset['symbol'], asset.get('asset_type','equity'), extra_texts)
        ctx_cache = {}
        SPECIALIST_ROLES = ('macro','sentiment','supply_chain','technical','geopolitical','sector','social_sentiment')

        def ctx(role):
            if role not in ctx_cache:
                base = serialize(ws, role=role, compact=False)
                ctx_cache[role] = f"{rel_ctx}\n\n{base}"
            return ctx_cache[role]

        # PHASE 1: ALL AGENTS WRITE TO GRAPH
        graph = SharedGraph(asset)
        agent_outputs = []
        prog(22, f'[Phase 1] Starting {len(p1)} specialist agents...')

        with ThreadPoolExecutor(max_workers=MAX_W) as ex:
            fut_map = {ex.submit(run_agent, ag,
                                 ctx(ag.role if ag.role in SPECIALIST_ROLES else 'macro'),
                                 graph): i
                       for i, ag in enumerate(p1)}
            done = 0
            for fut in as_completed(fut_map):
                ag = p1[fut_map[fut]]; done += 1
                pct = 22 + int(done/len(p1)*40)
                try:
                    out = fut.result(); agent_outputs.append(out)
                    prog(pct, f"[Phase 1] {ag.name}: {out['signal']} @{out['confidence']}% ({done}/{len(p1)})")
                except Exception as e:
                    log.error(f"Agent {ag.name} failed: {e}")
                    agent_outputs.append({
                        'role':'specialist','role_name':ag.name,'agent_id':ag.agent_id,
                        'agent_name':ag.name,'signal':'neutral','confidence':0,
                        'reasoning':str(e)[:200],'keyFactors':[],'butterflies':[],'edgeClaims':[],'_error':True})

        flow = graph.to_flow(); st = flow['stats']
        prog(63, f"[Graph] {st['total_nodes']} nodes | {st['total_edges']} edges | {st['bull']}B/{st['bear']}Be/{st['neut']}N")

        # PHASE 2: SYNTHESIZERS READ GRAPH
        synth_results = []
        for i, synth in enumerate(synths):
            pct = 65 + int(i/len(synths)*22)
            prog(pct, f"[Phase 2] {synth.name} reading complete graph...")
            try:
                synth_results.append(run_synthesizer(synth, graph))
            except Exception as e:
                log.error(f"Synthesizer {synth.name} failed: {e}")
                t = st['total_agents'] or 1
                synth_results.append({
                    'agent_id':synth.agent_id,'agent_name':synth.name,
                    'upProbability':round(st['bull']/t*100),'downProbability':round(st['bear']/t*100),
                    'neutralProbability':round(st['neut']/t*100),
                    'primaryDirection':'up' if st['bull']>st['bear'] else 'down' if st['bear']>st['bull'] else 'sideways',
                    'confidence':30,'bullCase':'','bearCase':'','keyRisks':['Synth failed'],
                    'topCatalysts':[],'topButterflyEffects':[],'socialAssessment':'N/A',
                    'worldStateHighlights':'N/A','signalConflicts':'N/A',
                    'technicalPicture':'N/A','fundamentalPicture':'N/A',
                    'summary':f"Fallback: {st['bull']}B/{st['bear']}Be/{st['neut']}N",'_fallback':True})

        # PHASE 3: SUPER SYNTHESIZER
        if super_a and synth_results:
            prog(88, f"[Phase 3] {super_a.name} reconciling {len(synth_results)} verdicts...")
            try: final = run_super_synthesizer(super_a, synth_results)
            except: final = synth_results[0]
        elif len(synth_results) > 1:
            up = round(sum(r['upProbability']   for r in synth_results)/len(synth_results))
            dn = round(sum(r['downProbability'] for r in synth_results)/len(synth_results))
            final = {**synth_results[0],'upProbability':up,'downProbability':dn,
                     'neutralProbability':100-up-dn,'_merged':len(synth_results)}
        else:
            final = synth_results[0] if synth_results else {}

        ws_ = ws or {}
        snapshot = {
            'fetched_at':     ws_.get('fetched_at'),
            'weather_alerts': [{'region':w['name'],'importance':w['importance'],'alerts':w['alerts']}
                                for w in ws_.get('weather',[]) if w.get('significant')],
            'commodities':    ws_.get('commodities',[])[:14],
            'options':        ws_.get('options'),
            'analyst':        ws_.get('analyst'),
            'earnings':       ws_.get('earnings'),
            'social_stats':   ws_.get('social',{}).get('stats'),
        }
        duration_ms = int((time.time()-t0)*1000)

        record = {
            'id': analysis_id, 'asset': asset, 'price': ws_.get('price'),
            'agent_outputs': agent_outputs, 'synthesizer_outputs': synth_results,
            'synthesis': final, 'graph': flow, 'data_snapshot': snapshot,
            'stats': {
                'phase1_agents': len(p1), 'synth_agents': len(synths),
                'super_synth': bool(super_a), 'total_nodes': st['total_nodes'],
                'total_edges': st['total_edges'], 'bull': st['bull'],
                'bear': st['bear'], 'neut': st['neut'],
                'social_total': social_total, 'weather_alerts': weather_alerts,
            },
            'created_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'duration_ms': duration_ms,
        }
        store.save_raw(analysis_id, record)

        _emit(task, status=TaskStatus.COMPLETED, progress=100,
              message='Analysis complete.',
              result={'analysis_id': analysis_id, 'symbol': asset['symbol'],
                      'direction': final.get('primaryDirection'),
                      'up': final.get('upProbability'),
                      'down': final.get('downProbability'),
                      'confidence': final.get('confidence'),
                      'analysis': record})

        if on_complete:
            try: on_complete(analysis_id, record)
            except Exception as e: log.warning(f'on_complete error: {e}')

        log.info(f"Pipeline done: {analysis_id} in {duration_ms}ms")

    except Exception as e:
        import traceback as tb_mod
        tb = tb_mod.format_exc()
        log.error(f"Pipeline failed: {e}\n{tb}")
        _emit(task, status=TaskStatus.FAILED, progress=0,
              message=f'Failed: {str(e)}', error=tb)

File: backend/agents/orchestrator.js
const { v4: uuidv4 } = require('uuid');
const { fetchPrice, fetchHistory, fetchMacroContext } = require('../services/market');
const { buildWorldState, worldStateToText } = require('../services/worldstate');
const { fetchAllSocial } = require('../services/social');
const { runAgentWriteToGraph, runGraphSynthesizer } = require('../agents/runner');
const { SharedGraph } = require('../agents/graph');
const { saveAnalysis } = require('../services/storage');

const CONCURRENCY = 5;

async function runConcurrent(tasks, limit) {
  const results = new Array(tasks.length);
  let i = 0;
  const workers = Array.from({ length: Math.min(limit, tasks.length) }, async () => {
    while (i < tasks.length) {
      const idx = i++;
      try { results[idx] = { status: 'fulfilled', value: await tasks[idx]() }; }
      catch (err) { results[idx] = { status: 'rejected', reason: err }; }
    }
  });
  await Promise.all(workers);
  return results;
}

/**
 * THE FULL TWO-PHASE PIPELINE
 *
 * PRE-FETCH (parallel, once):
 *   - Market data: price, OHLCV history, financial news
 *   - World state: weather (8 regions), commodity prices (10), shipping BDI,
 *                  conflicts, agriculture, energy, central banks, trade, economic
 *   - Social data: Reddit (5 subs), X/Nitter, StockTwits, HackerNews
 *                  — each post manipulation-scored before agents see it
 *
 * PHASE 1: All specialist agents run concurrently (up to CONCURRENCY at a time).
 *          Each receives identical market + world + social context.
 *          Each writes independently to the SharedGraph.
 *          No agent sees another agent's output.
 *
 * PHASE 2: One synthesizer receives the complete assembled graph as its only input.
 *          Produces final probability verdict + analysis.
 */
async function runFullAnalysis(asset, agents, onProgress = () => {}) {
  const analysisId = uuidv4();
  const startTime = Date.now();

  const enabledAgents = agents.filter(a => a.enabled !== false);
  if (enabledAgents.length === 0) throw new Error('No enabled agents. Add at least one in the Agents tab.');

  const synthAgents  = enabledAgents.filter(a => a.role === 'synthesizer');
  const writingAgents = enabledAgents.filter(a => a.role !== 'synthesizer');
  const synthAgent   = synthAgents[0] || enabledAgents[0];
  const phase1Agents = writingAgents.length > 0 ? writingAgents : enabledAgents;

  const totalSteps = 7 + phase1Agents.length + 2;

  // ── PRE-FETCH: market data ────────────────────────────────────────
  onProgress(1, totalSteps, 'Fetching current price...');
  const price = await fetchPrice(asset.symbol, asset.alphaVantageKey).catch(() => null);

  onProgress(2, totalSteps, 'Fetching price history...');
  const history = await fetchHistory(asset.symbol, 30, asset.alphaVantageKey).catch(() => []);

  onProgress(3, totalSteps, 'Fetching financial news...');
  const { assetNews, globalNews } = await fetchMacroContext(asset.symbol).catch(() => ({ assetNews: [], globalNews: [] }));
  const allNews = [...(assetNews || []), ...(globalNews || [])];

  // ── PRE-FETCH: world state ────────────────────────────────────────
  onProgress(4, totalSteps, 'Building world state (weather, commodities, conflicts, shipping)...');
  const worldState = await buildWorldState(asset.symbol).catch(() => null);
  const worldStateText = worldState ? worldStateToText(worldState) : '';

  // ── PRE-FETCH: social media ───────────────────────────────────────
  onProgress(5, totalSteps, 'Fetching social media signals (Reddit, X, StockTwits, HackerNews)...');
  const socialData = await fetchAllSocial(asset.symbol).catch(() => null);

  onProgress(6, totalSteps, `Data ready. World: ${worldState ? 'ok' : 'partial'} | Social: ${socialData?.stats?.total || 0} posts (${socialData?.stats?.suspicious || 0} flagged)`);

  // ── PHASE 1: All agents write to shared graph ─────────────────────
  const graph = new SharedGraph(asset);
  const agentOutputs = [];
  let step = 7;

  const tasks = phase1Agents.map((agent, idx) => async () => {
    onProgress(step + idx, totalSteps, `[Phase 1] "${agent.name}" analyzing → graph... (${idx + 1}/${phase1Agents.length})`);
    return runAgentWriteToGraph(agent, asset, price, history, allNews, graph, worldStateText, socialData);
  });

  const results = await runConcurrent(tasks, CONCURRENCY);

  results.forEach((r, idx) => {
    if (r.status === 'fulfilled') {
      agentOutputs.push(r.value);
    } else {
      const agent = phase1Agents[idx];
      const stub = { signal: 'neutral', confidence: 0, reasoning: `Failed: ${r.reason?.message || 'unknown'}`, keyFactors: [], butterflies: [], error: true };
      graph.addSignal({ agentId: agent.id, agentName: agent.name, role: agent.role || 'specialist', roleName: agent.name, ...stub });
      agentOutputs.push({ agentId: agent.id, agentName: agent.name, roleName: agent.name, nodeId: null, ...stub });
    }
  });

  step += phase1Agents.length;

  // ── Graph assembled ───────────────────────────────────────────────
  const flowGraph = graph.toFlowFormat();
  onProgress(step++, totalSteps,
    `[Graph complete] ${flowGraph.stats.totalNodes} nodes | ${flowGraph.stats.totalEdges} edges | ` +
    `${flowGraph.stats.bull}B / ${flowGraph.stats.bear}Be / ${flowGraph.stats.neut}N`);

  // ── PHASE 2: Synthesizer reads the full graph ─────────────────────
  onProgress(step++, totalSteps, `[Phase 2] Synthesizer "${synthAgent.name}" reading complete graph...`);

  let synthesis;
  try {
    synthesis = await runGraphSynthesizer(synthAgent, asset, graph);
  } catch (err) {
    const t = flowGraph.stats.totalAgents || 1;
    synthesis = {
      upProbability: Math.round((flowGraph.stats.bull / t) * 100),
      downProbability: Math.round((flowGraph.stats.bear / t) * 100),
      neutralProbability: Math.round((flowGraph.stats.neut / t) * 100),
      primaryDirection: flowGraph.stats.bull > flowGraph.stats.bear ? 'up' : flowGraph.stats.bear > flowGraph.stats.bull ? 'down' : 'sideways',
      confidence: 30,
      bullCase: `${flowGraph.stats.bull}/${t} agents bullish.`,
      bearCase: `${flowGraph.stats.bear}/${t} agents bearish.`,
      keyRisks: ['Synthesizer failed -- vote aggregation used'],
      topCatalysts: [], topButterflyEffects: [],
      socialSignalAssessment: 'N/A', worldStateHighlights: 'N/A', signalConflicts: 'N/A',
      summary: `Fallback: ${flowGraph.stats.bull}B/${flowGraph.stats.bear}Be/${flowGraph.stats.neut}N from ${t} agents.`,
      fallback: true, error: err.message,
    };
  }

  const analysis = {
    id: analysisId,
    asset,
    price,
    agentOutputs,
    graph: flowGraph,
    synthesis,
    worldState: {
      fetchedAt: worldState?.fetchedAt,
      weatherAlerts: (worldState?.weather || []).filter(w => w.significant).map(w => ({ region: w.region, alerts: w.alerts })),
      commodities: worldState?.commodities || [],
      shippingBDI: worldState?.shipping?.bdi || null,
    },
    social: socialData ? {
      stats: socialData.stats,
      topCleanPosts: (socialData.clean || []).slice(0, 5).map(p => ({ platform: p.platform, title: p.title, score: p.score })),
    } : null,
    stats: {
      totalAgents: phase1Agents.length,
      totalNodes: flowGraph.stats.totalNodes,
      totalEdges: flowGraph.stats.totalEdges,
      bull: flowGraph.stats.bull,
      bear: flowGraph.stats.bear,
      neut: flowGraph.stats.neut,
      synthAgent: synthAgent.name,
      socialPostsTotal: socialData?.stats?.total || 0,
      socialPostsSuspicious: socialData?.stats?.suspicious || 0,
      weatherAlertsCount: (worldState?.weather || []).filter(w => w.significant).length,
    },
    createdAt: new Date().toISOString(),
    durationMs: Date.now() - startTime,
  };

  await saveAnalysis(analysis);
  return analysis;
}

module.exports = { runFullAnalysis };
